testing/test_pytester.py

```diff
@@ -542,17 +542,27 @@ def test_no_matching(function):
     func(bad_pattern)  # bad pattern does not match any line: passes
 
 
-def test_pytester_addopts(request, monkeypatch):
+def test_pytester_addopts_before_testdir(request, monkeypatch):
+    orig = os.environ.get("PYTEST_ADDOPTS", None)
     monkeypatch.setenv("PYTEST_ADDOPTS", "--orig-unused")
-
     testdir = request.getfixturevalue("testdir")
+    assert "PYTEST_ADDOPTS" not in os.environ
+    testdir.finalize()
+    assert os.environ.get("PYTEST_ADDOPTS") == orig
 
-    try:
-        assert "PYTEST_ADDOPTS" not in os.environ
-    finally:
-        testdir.finalize()
 
-    assert os.environ["PYTEST_ADDOPTS"] == "--orig-unused"
+@pytest.mark.parametrize("method", ("setenv", "delenv"))
+def test_testdir_respects_monkeypatch(method, testdir, monkeypatch):
```

Comment:
Regardless if we follow my previous suggestion about using `monkeypatch` during `Testdir.__init__`, can we have a more user facing test here? The test I posted also works on your branch, and tests the actual behavior we want:

```python
@pytest.mark.parametrize("method", ("setenv", "delenv"))
def test_testdir_respects_monkeypatch(method, testdir, monkeypatch):
    testdir.makepyfile("""
        import os
        def test_home():
            print("HOME:", os.environ.get("HOME"))
    """)

    if method == "setenv":
        monkeypatch.setenv("HOME", "anotherhome")
    else:
        assert method == "delenv"
        monkeypatch.delenv("HOME", raising=False)

    result = testdir.runpytest("-vs")

    if method == "setenv":
        result.stdout.fnmatch_lines("*HOME: anotherhome*")
    else:
        assert method == "delenv"
        result.stdout.fnmatch_lines("*HOME: None*")
```

Reply:
Not strictly opposed - my initial test was meant to be a unit test for `_get_env_run_update` (which would be gone then though).  It's much faster though (1.2s for 100 runs (200 passes), compared to 20s for the above).

Reply:
Well I fear that by testing the inner mechanism and not testing the expected behavior, we might introduce a regression later and fail to notice. But I will leave this one to you.

Reply:
I think this is covered by existing/changed tests.